include/cuco/detail/open_addressing_ref_impl.cuh

```diff
@@ -0,0 +1,619 @@
+   */
+  template <typename ProbeKey, typename Predicate>
+  [[nodiscard]] __device__ const_iterator
+  find(cooperative_groups::thread_block_tile<cg_size> const& group,
+       ProbeKey const& key,
+       Predicate const& predicate) const noexcept
+  {
+    auto probing_iter = probing_scheme_(group, key, storage_ref_.num_windows());
+
+    while (true) {
+      auto const window_slots = storage_ref_[*probing_iter];
+
+      auto const [state, intra_window_index] = [&]() {
+        for (auto i = 0; i < window_size; ++i) {
+          switch (predicate(window_slots[i], key)) {
+            case detail::equal_result::EMPTY: return window_results{detail::equal_result::EMPTY, i};
+            case detail::equal_result::EQUAL: return window_results{detail::equal_result::EQUAL, i};
+            default: continue;
+          }
+        }
+        // returns dummy index `-1` for UNEQUAL
+        return window_results{detail::equal_result::UNEQUAL, -1};
+      }();
+
+      // Find a match for the probe key, thus return an iterator to the entry
+      auto const group_finds_match = group.ballot(state == detail::equal_result::EQUAL);
+      if (group_finds_match) {
+        auto const src_lane = __ffs(group_finds_match) - 1;
+        auto const res      = group.shfl(
+          reinterpret_cast<intptr_t>(&(*(storage_ref_.data() + *probing_iter))[intra_window_index]),
+          src_lane);
+        return const_iterator{reinterpret_cast<value_type*>(res)};
+      }
+
+      // Find an empty slot, meaning that the probe key isn't present in the container
+      if (group.any(state == detail::equal_result::EMPTY)) { return this->end(); }
+
+      ++probing_iter;
+    }
+  }
+
+ private:
+  /// Three-way insert result enum
+  enum class insert_result : int32_t { CONTINUE = 0, SUCCESS = 1, DUPLICATE = 2 };
+
+  /**
+   * @brief Helper struct to store intermediate window probing results.
+   */
+  struct window_results {
+    detail::equal_result state_;  ///< Equal result
+    int32_t intra_window_index_;  ///< Intra-window index
+
+    /**
+     * @brief Constructs window_results.
+     *
+     * @param state The three way equality result
+     *@param Intra-window index
+     */
+    __device__ explicit constexpr window_results(detail::equal_result state, int32_t index) noexcept
+      : state_{state}, intra_window_index_{index}
+    {
+    }
+  };
+
+  /**
+   * @brief Attempts to insert an element into a slot.
+   *
+   * @note Dispatches the correct implementation depending on the container
+   * type and presence of other operator mixins.
+   *
+   * @tparam Predicate Predicate type
+   *
+   * @param slot Pointer to the slot in memory
+   * @param value Element to insert
+   * @param predicate Predicate used to compare slot content against `key`
+   *
+   * @return Result of this operation, i.e., success/continue/duplicate
+   */
+  template <typename Predicate>
+  [[nodiscard]] __device__ insert_result attempt_insert(value_type* slot,
```

Comment:
Should this function also implement the b2b CAS idea or where shall we put it?

Reply:
would be in a followup PR